contracts/WETH10.sol

```diff
@@ -1,102 +1,106 @@
+    mapping (address => uint256)                       public  balanceOf;
+    mapping (address => mapping (address => uint256))  public  allowance;
+    mapping (address => uint256)                       public  nonces;
     
-    constructor() public {
-        name = "Wrapped Ether";
-        symbol = "WETH";
-        decimals = 18;
-        uint chainId;
+    constructor() {
+        uint256 chainId;
         assembly {chainId := chainid()}
         DOMAIN_SEPARATOR = keccak256(
             abi.encode(
-                keccak256("EIP712Domain(string name,string version,uint chainId,address verifyingContract)"),
+                keccak256("EIP712Domain(string name,string version,uint256 chainId,address verifyingContract)"),
                 keccak256(bytes(name)),
                 keccak256(bytes("1")),
                 chainId,
                 address(this)));
     }
 
     receive() external payable {
-        deposit();
+        balanceOf[msg.sender] += msg.value;
+        emit Transfer(address(0), msg.sender, msg.value);
     }
     
-    function deposit() public payable {
+    function deposit() external payable {
         balanceOf[msg.sender] += msg.value;
-        emit Deposit(msg.sender, msg.value);
+        emit Transfer(address(0), msg.sender, msg.value);
     }
     
-    function withdraw(uint wad) external {
-        require(balanceOf[msg.sender] >= wad, "!balance");
-        balanceOf[msg.sender] -= wad;
-        (bool success, ) = msg.sender.call.value(wad)("");
+    function withdraw(uint256 value) external {
+        uint256 balance = balanceOf[msg.sender];
+        require(balance >= value, "!balance");
+        balance -= value;
+        (bool success, ) = msg.sender.call{value: value}("");
         require(success, "!withdraw");
-        emit Withdrawal(msg.sender, wad);
+        emit Transfer(msg.sender, address(0), value);
     }
 
-    function totalSupply() external view returns (uint) {
+    function totalSupply() external view returns (uint256) {
         return address(this).balance;
     }
     
-    function _approve(address src, address guy, uint wad) internal {
-        allowance[src][guy] = wad;
-        emit Approval(src, guy, wad);
+    function _approve(address from, address spender, uint256 value) internal {
+        allowance[from][spender] = value;
+        emit Approval(from, spender, value);
     }
     
-    function approve(address guy, uint wad) external returns (bool) {
-        _approve(msg.sender, guy, wad); 
+    function approve(address spender, uint256 value) external returns (bool) {
+        _approve(msg.sender, spender, value); 
         return true;
     }
     
-    function transfer(address dst, uint wad) external returns (bool) {
-        return transferFrom(msg.sender, dst, wad);
+    function transfer(address to, uint256 value) external returns (bool) {
+        uint256 balance = balanceOf[msg.sender];
+        require(balance >= value, "!balance");
+
+        balance -= value;
```

Comment:
Here you are only updating the local copy of the balance. The original implementation was right.

Reply:
Original implementation uses 2 SLOADs instead of 1, it should be this:

```
        uint256 balance = balanceOf[msg.sender];
        require(balance >= value, "!balance");
        balanceOf[msg.sender] -= value;
```

Reply:
the additional sload should be optimized away with the optimizer is activated

Reply:
great catches 🙌 !! will tweak (afk)